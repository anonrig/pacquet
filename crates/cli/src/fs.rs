use std::{io, os, path::Path};

#[cfg(unix)]
pub fn symlink_dir<P: AsRef<Path>>(original: P, link: P) -> io::Result<()> {
    os::unix::fs::symlink(original, link)
}

#[cfg(windows)]
pub fn symlink_dir<P: AsRef<Path>>(original: P, link: P) -> io::Result<()> {
    os::windows::fs::symlink_dir(original, link)
}

#[cfg(test)]
pub fn get_filenames_in_folder(path: &Path) -> Vec<String> {
    let mut files = std::fs::read_dir(path)
        .unwrap()
        .map(|e| e.unwrap().file_name().to_string_lossy().to_string())
        .collect::<Vec<_>>();

    files.sort();
    files
}

#[cfg(test)]
pub fn get_all_folders(root: &std::path::PathBuf) -> Vec<String> {
    let mut files = Vec::new();
    for entry in walkdir::WalkDir::new(root) {
        let entry = entry.unwrap();
        let entry_path = entry.path().to_path_buf();
        if entry.file_type().is_dir() || entry.file_type().is_symlink() {
            // We need this mutation to ensure that both Unix and Windows paths resolves the same.
            // TODO: Find a better way to do this?
            let simple_path = entry_path
                .strip_prefix(root)
                .unwrap()
                .components()
                .map(|c| c.as_os_str().to_string_lossy().to_string())
                .collect::<Vec<_>>()
                .join("/");

            if !simple_path.is_empty() {
                files.push(simple_path.as_os_string());
            }
        }
    }
    files.sort();
    files
}


