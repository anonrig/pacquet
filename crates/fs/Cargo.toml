[package]
name                 = "pacquet-fs"
description          = "Filesystem utility functions used by pacquet"
version              = "0.0.1"
publish              = false
authors.workspace    = true
edition.workspace    = true
homepage.workspace   = true
keywords.workspace   = true
license.workspace    = true
repository.workspace = true

[dependencies]
pacquet-task-queue = { workspace = true }

derive_more = { workspace = true }
miette      = { workspace = true }

[target.'cfg(windows)'.dependencies]
junction = { workspace = true }
