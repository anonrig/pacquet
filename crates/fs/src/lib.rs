mod ensure_file;
mod symlink_dir;

pub use ensure_file::*;
pub use symlink_dir::*;

pub mod file_mode;
