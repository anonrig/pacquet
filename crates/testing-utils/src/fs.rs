use std::{io, path::Path};

pub fn get_filenames_in_folder(path: &Path) -> Vec<String> {
    let mut files = std::fs::read_dir(path)
        .unwrap()
        .map(|e| e.unwrap().file_name().to_string_lossy().to_string())
        .collect::<Vec<_>>();

    files.sort();
    files
}

pub fn get_all_folders(root: &std::path::Path) -> Vec<String> {
    let mut files = Vec::new();
    for entry in walkdir::WalkDir::new(root) {
        let entry = entry.unwrap();
        let entry_path = entry.path();
        if entry.file_type().is_dir() || entry.file_type().is_symlink() {
            // We need this mutation to ensure that both Unix and Windows paths resolves the same.
            // TODO: Find a better way to do this?
            let simple_path = entry_path
                .strip_prefix(root)
                .unwrap()
                .components()
                .map(|c| c.as_os_str().to_str().expect("invalid UTF-8"))
                .collect::<Vec<_>>()
                .join("/");

            if !simple_path.is_empty() {
                files.push(simple_path);
            }
        }
    }
    files.sort();
    files
}

// Helper function to check if a path is a symlink or junction
pub fn is_symlink_or_junction(path: &Path) -> io::Result<bool> {
    #[cfg(windows)]
    return junction::exists(&path);

    #[cfg(not(windows))]
    return Ok(path.is_symlink());
}
