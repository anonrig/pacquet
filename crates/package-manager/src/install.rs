use crate::{InstallFrozenLockfile, InstallWithoutLockfile};
use pacquet_lockfile::Lockfile;
use pacquet_npmrc::Npmrc;
use pacquet_package_manifest::{DependencyGroup, PackageManifest};
use pacquet_tarball::MemCache;
use reqwest::Client;

/// This subroutine does everything `pacquet install` is supposed to do.
#[must_use]
pub struct Install<'a, DependencyGroupList>
where
    DependencyGroupList: IntoIterator<Item = DependencyGroup>,
{
    pub tarball_mem_cache: &'a MemCache,
    pub http_client: &'a Client,
    pub config: &'static Npmrc,
    pub manifest: &'a PackageManifest,
    pub lockfile: Option<&'a Lockfile>,
    pub dependency_groups: DependencyGroupList,
    pub frozen_lockfile: bool,
}

impl<'a, DependencyGroupList> Install<'a, DependencyGroupList>
where
    DependencyGroupList: IntoIterator<Item = DependencyGroup>,
{
    /// Execute the subroutine.
    pub async fn run(self) {
        let Install {
            tarball_mem_cache,
            http_client,
            config,
            manifest,
            lockfile,
            dependency_groups,
            frozen_lockfile,
        } = self;

        tracing::info!(target: "pacquet::install", "Start all");

        match (config.lockfile, frozen_lockfile, lockfile) {
            (false, _, _) => {
                InstallWithoutLockfile {
                    tarball_mem_cache,
                    http_client,
                    config,
                    manifest,
                    dependency_groups,
                }
                .run()
                .await;
            }
            (true, false, Some(_)) | (true, false, None) | (true, true, None) => {
                unimplemented!();
            }
            (true, true, Some(lockfile)) => {
                let Lockfile { lockfile_version, project_snapshot, packages, .. } = lockfile;
                assert_eq!(lockfile_version.major, 6); // compatibility check already happens at serde, but this still helps preventing programmer mistakes.

                InstallFrozenLockfile {
                    http_client,
                    config,
                    project_snapshot,
                    packages: packages.as_ref(),
                    dependency_groups,
                }
                .run()
                .await;
            }
        }

        tracing::info!(target: "pacquet::install", "Complete all");
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use pacquet_npmrc::Npmrc;
    use pacquet_package_manifest::{DependencyGroup, PackageManifest};
    use pacquet_testing_utils::fs::{get_all_folders, is_symlink_or_junction};
    use std::env;
    use tempfile::tempdir;

    #[tokio::test]
    async fn should_install_dependencies() {
        let dir = tempdir().unwrap();
        let store_dir = dir.path().join("pacquet-store");
        let project_root = dir.path().join("project");
        let modules_dir = project_root.join("node_modules"); // TODO: we shouldn't have to define this
        let virtual_store_dir = modules_dir.join(".pacquet"); // TODO: we shouldn't have to define this

        let manifest_path = dir.path().join("package.json");
        let mut manifest = PackageManifest::create_if_needed(manifest_path.clone()).unwrap();

        manifest.add_dependency("is-odd", "3.0.1", DependencyGroup::Prod).unwrap();
        manifest
            .add_dependency("fast-decode-uri-component", "1.0.1", DependencyGroup::Dev)
            .unwrap();

        manifest.save().unwrap();

        let mut config = Npmrc::new();
        config.store_dir = store_dir.into();
        config.modules_dir = modules_dir.to_path_buf();
        config.virtual_store_dir = virtual_store_dir.to_path_buf();
        let config = config.leak();

        Install {
            tarball_mem_cache: &Default::default(),
            http_client: &Default::default(),
            config,
            manifest: &manifest,
            lockfile: None,
            dependency_groups: [
                DependencyGroup::Prod,
                DependencyGroup::Dev,
                DependencyGroup::Optional,
            ],
            frozen_lockfile: false,
        }
        .run()
        .await;

        // Make sure the package is installed
        assert!(is_symlink_or_junction(&project_root.join("node_modules/is-odd")).unwrap());
        assert!(project_root.join("node_modules/.pacquet/is-odd@3.0.1").exists());
        // Make sure it installs direct dependencies
        assert!(!project_root.join("node_modules/is-number").exists());
        assert!(project_root.join("node_modules/.pacquet/is-number@6.0.0").exists());
        // Make sure we install dev-dependencies as well
        assert!(is_symlink_or_junction(
            &project_root.join("node_modules/fast-decode-uri-component")
        )
        .unwrap());
        assert!(project_root
            .join("node_modules/.pacquet/fast-decode-uri-component@1.0.1")
            .is_dir());

        insta::assert_debug_snapshot!(get_all_folders(&project_root));
    }
}
