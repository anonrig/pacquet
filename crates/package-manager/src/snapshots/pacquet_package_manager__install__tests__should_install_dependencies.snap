---
source: crates/package-manager/src/install.rs
assertion_line: 140
expression: get_all_folders(&project_root)
---
[
    "node_modules",
    "node_modules/.pacquet",
    "node_modules/.pacquet/@pnpm+x@1.0.0",
    "node_modules/.pacquet/@pnpm+x@1.0.0/node_modules",
    "node_modules/.pacquet/@pnpm+x@1.0.0/node_modules/@pnpm",
    "node_modules/.pacquet/@pnpm+x@1.0.0/node_modules/@pnpm/x",
    "node_modules/.pacquet/@pnpm+xyz@1.0.0",
    "node_modules/.pacquet/@pnpm+xyz@1.0.0/node_modules",
    "node_modules/.pacquet/@pnpm+xyz@1.0.0/node_modules/@pnpm",
    "node_modules/.pacquet/@pnpm+xyz@1.0.0/node_modules/@pnpm/x",
    "node_modules/.pacquet/@pnpm+xyz@1.0.0/node_modules/@pnpm/xyz",
    "node_modules/.pacquet/@pnpm+xyz@1.0.0/node_modules/@pnpm/y",
    "node_modules/.pacquet/@pnpm+xyz@1.0.0/node_modules/@pnpm/z",
    "node_modules/.pacquet/@pnpm+y@1.0.0",
    "node_modules/.pacquet/@pnpm+y@1.0.0/node_modules",
    "node_modules/.pacquet/@pnpm+y@1.0.0/node_modules/@pnpm",
    "node_modules/.pacquet/@pnpm+y@1.0.0/node_modules/@pnpm/y",
    "node_modules/.pacquet/@pnpm+z@1.0.0",
    "node_modules/.pacquet/@pnpm+z@1.0.0/node_modules",
    "node_modules/.pacquet/@pnpm+z@1.0.0/node_modules/@pnpm",
    "node_modules/.pacquet/@pnpm+z@1.0.0/node_modules/@pnpm/z",
    "node_modules/.pacquet/@pnpm.e2e+hello-world-js-bin@1.0.0",
    "node_modules/.pacquet/@pnpm.e2e+hello-world-js-bin@1.0.0/node_modules",
    "node_modules/.pacquet/@pnpm.e2e+hello-world-js-bin@1.0.0/node_modules/@pnpm.e2e",
    "node_modules/.pacquet/@pnpm.e2e+hello-world-js-bin@1.0.0/node_modules/@pnpm.e2e/hello-world-js-bin",
    "node_modules/@pnpm",
    "node_modules/@pnpm/xyz",
    "node_modules/@pnpm.e2e",
    "node_modules/@pnpm.e2e/hello-world-js-bin",
]
